File: spring-boot-domain-persistence-modeling-java/src/main/java/x/scratch/child/ChildDetails.java
package x.scratch.child;

import javax.annotation.Nonnull;
import java.util.Set;

public interface ChildDetails
        extends Comparable<ChildDetails> {
    @Nonnull
    String getNaturalId();

    String getParentNaturalId();

    String getValue();

    @Nonnull
    Set<String> getSubchildren();

    int getVersion();

    @Override
    default int compareTo(@Nonnull final ChildDetails that) {
        return getNaturalId().compareTo(that.getNaturalId());
    }
}

File: spring-boot-domain-persistence-modeling-kotlin/src/main/kotlin/x/domainpersistencemodeling/java/ChildDetails.kt
package x.domainpersistencemodeling.java

interface ChildDetails : Comparable<Child> {
    val naturalId: String
    val parentNaturalId: String?
    val value: String?
    val subchildren: Set<String>
    val version: Int

    override operator fun compareTo(other: Child): Int {
        return naturalId.compareTo(other.naturalId)
    }
}
